### PR TITLE
Conditional dependencies in lockdir

### PR DESCRIPTION
**Pull request by gridbugs**
This changes lockdir generation so that dependencies which are conditional on the variables `:with-test` and `:with-doc` are added to the lockdir.

### CODE REVIEW

File: src/dune_pkg/opam_solver.ml
Comment (Leonidas-from-XIV) [inline src/dune_pkg/opam_solver.ml:109]: Isn't this the env function from the solver? Maybe it can be reused?

File: src/dune_pkg/opam_solver.ml
Comment (Leonidas-from-XIV) [inline src/dune_pkg/opam_solver.ml:163]: It should probably not warn to the Opam console.

**Review by Leonidas-from-XIV — COMMENTED** [review]
Looks good and a very useful change. Some nitpicking on the code from the 0install solver.

File: src/dune_pkg/opam_solver.ml
Comment (gridbugs) [reply src/dune_pkg/opam_solver.ml:83]: It means that dependencies marked as "post" will be resolved. I've never encountered this variable in the wild but it tells opam that the dependency can be compiled after the current package. https://opam.ocaml.org/doc/Manual.html#pkgvar-post I set it to true because that's what `Opam_0install.Dir_context` does and it seems like the most useful default setting. Perhaps one day we'll need to add support to dune to set it to false to deal with dependency cycles.

File: src/dune_pkg/opam_solver.ml
Comment (gridbugs) [reply src/dune_pkg/opam_solver.ml:109]: Actually yes, I've since dug a bit deeper into opam filters in another change I'm working on and I think we can simplify this by resolving the `test` and `doc` variables to true before calling the base context's `filter_deps` on the result. Will update.

File: src/dune_pkg/opam_solver.ml
Comment (gridbugs) [reply src/dune_pkg/opam_solver.ml:163]: Oops yes you're right.

File: src/dune_pkg/opam_solver.ml
Comment (Leonidas-from-XIV) [reply src/dune_pkg/opam_solver.ml:83]: Thanks for the explanation, I have never encountered these before. Out of curiosity I checked opam-repository and there are two non-compiler packages (only 2 versions overall) that use these: `sail.0.15` and `ctypes-foreign.0.18.0` so indeed these don't seem to be used much.

**Comment by gridbugs** [comment]
I removed all the code borrowed/modified from opam_0install and instead added a module `Filtered_formula` with functions for transforming opam formulas and now `filter_deps` just uses that to replace `with-test` and `with-doc` with `true` before calling the base context's `filter_deps` on the result. This also means we could get rid of some of the `*_for_dune` modules previously in this change.

Also in the process I noticed that we were setting the `with-dev-setup` variable which I don't think we want so now we just set the `with-test` and `with-doc` variables to true.

File: src/dune_pkg/opam_solver.ml
Comment (Leonidas-from-XIV) [inline src/dune_pkg/opam_solver.ml:51]: Maybe `variable_name` would be better as an `OpamVariable`, then one could use its `equal` function, thus making the named arguments less important.

File: src/dune_pkg/opam_solver.ml
Comment (Leonidas-from-XIV) [inline src/dune_pkg/opam_solver.ml:45]: Cool, didn't know these functions. `map_filters_bottom_up` seems quite useful.

Comment (gridbugs) [reply src/dune_pkg/opam_solver.ml:45]: Yeh `OpamFilter.map_up` is very handy. I just learnt about it recently too.

**Comment by rgrinberg** [comment]
A couple of things that are missing from this PR:

* The ability to customize this behavior. In particular, it should be possible not to have docs or test deps. Also, it should be possible to include one of the two.
* The second thing we're missing is a way to propagate which deps were added because of the deps or tests flags. This is so that installing the software can use the same lock file (when applicable).

The first one needs to be done in this PR, while the second one can be done in subsequent work.

**Comment by gridbugs** [comment]
>     * The second thing we're missing is a way to propagate which deps were added because of the deps or tests flags. This is so that installing the software can use the same lock file (when applicable).

Can you clarify this point. Do you mean including in package metadata whether the package was included in the lockdir because it's a test/doc dependency so that we can omit those packages when using the lockdir for installation?

**Comment by gridbugs** [comment]
Cool makes sense. I'll update this PR so that `with-test` and `with-doc` can be configured per build context and make a new PR which tracks which packages were included due to these flags in a later change.

**Comment by gridbugs** [comment]
@rgrinberg I've added a `Solver_env.t` type which stores variables used by the solver. Currently it just stores the flags `:with-test` and `:with-doc` but I think it would make sense to use this for all opam variables (e.g. system variables) and replace the `Opam.Env.t` type completely. I'm experimenting with that on another branch but I won't add it to this PR so it doesn't get too complicated.

File: src/dune_pkg/opam_solver.ml
Comment (rgrinberg) [inline src/dune_pkg/opam_solver.ml:60]: You could just add a fold function to `Solver_env.Flag.Set`. It doesn't really matter much, but why favor list conversion when all you need is a fold?

File: src/dune_pkg/opam_solver.mli
Comment (rgrinberg) [inline src/dune_pkg/opam_solver.mli:18]: Don't we have a version preference type somewhere that we can use?

File: src/dune_pkg/opam_solver.mli
Comment (rgrinberg) [inline src/dune_pkg/opam_solver.mli:8]: Not entirely related, but I would prefer this naked `string` type turned into something abstract.

**Review by rgrinberg — APPROVED** [review]

File: src/dune_pkg/opam_solver.mli
Comment (gridbugs) [reply src/dune_pkg/opam_solver.mli:8]: I'm planning to remove this argument soon when I move all environment variables inside `Solver_env.t` so I'll leave it for now.

**Comment by gridbugs** [comment]
@rgrinberg does anything more need to happen before this can be merged?

**Comment by rgrinberg** [comment]
Ready from me. Waiting on @Leonidas-from-XIV to review

File: test/blackbox-tests/test-cases/pkg/solver-flags-per-context.t
Comment (Leonidas-from-XIV) [inline test/blackbox-tests/test-cases/pkg/solver-flags-per-context.t:52]: I wonder whether it would make sense to implement the set language so `(flags :standard)` can be specified.

**Review by Leonidas-from-XIV — APPROVED** [review]
I think it looks quite good, some minor comments when I looked over the diff.

File: test/blackbox-tests/test-cases/pkg/solver-flags-per-context.t
Comment (gridbugs) [reply test/blackbox-tests/test-cases/pkg/solver-flags-per-context.t:52]: Ah yes I like the idea of allowing `:standard` here. It won't be hard to add as long as we don't allow `:include` which I think is reasonable.

Comment (gridbugs) [reply test/blackbox-tests/test-cases/pkg/solver-flags-per-context.t:52]: Updated to use the ordered set language for this field and added a test using the `:standard` keyword.

**Comment by gridbugs** [comment]
> Yes, that's what I meant. Otherwise, we'll have to always build all the test deps.

@rgrinberg on the topic of storing additional metadata to keep track of which dependencies are test/doc-only so we don't build test/doc deps when installing, I thought about this some more and I don't think this is necessary. The solver flags apply only to dependencies of local packages define in `dune-project` and the dependency specification in `dune-project` already stores whether a package is a test or doc dependency.

**Pull request merged**
